File: 784.字母大小写全排列/letterCasePermutation.go
package problem0784

// Tips: 位运算转换字母大小写, 大小写字母二进制表示值相差32
// A(65): 0100 0001
// a(97): 0110 0001
// 可以观察出 A ^ 0010 0000 = a     a ^ 0010 0000 = A
// 回溯解决
func letterCasePermutation(S string) []string {
	var result []string
	backTracking(&result, []byte(S), 0)
	return result
}

func backTracking(result *[]string, s []byte, index int) {
	if index == len(s) {
		*result = append(*result, string(s))
		return
	}
	backTracking(result, s, index+1)
	if s[index] >= 65 {
		s[index] ^= 32
		backTracking(result, s, index+1)
	}
}

File: 784.字母大小写全排列/letterCasePermutation.js
// 简单回溯
function letterCasePermutation(s) {
	const result = []
	const n = s.length
	const backTrack = (path, i) => {
		if (i === n) {
			result.push(path)
			return
		}
		if ('a' <= s[i] && s[i] <= 'z') {
			backTrack(path + s[i].toUpperCase(), i + 1)
		} else if ('A' <= s[i] && s[i] <= 'Z') {
			backTrack(path + s[i].toLowerCase(), i + 1)
		}
		backTrack(path + s[i], i + 1)
	}
	backTrack('', 0)
	return result
}